File: 08-handheldHalting/python/partOne.py
from common import *


def partOne(instr: str) -> int:
    instructions = parse(instr)

    acc = 0
    pc = 0

    visited = []  # indexes of visited instructions

    while True:

        if pc in visited:
            return acc
        else:
            visited.append(pc)

        cir = instructions[pc]

        if cir.opcode == "jmp":
            pc += cir.operand
        else:
            if cir.opcode == "acc":
                acc += cir.operand
            elif cir.opcode == "nop":
                pass
            pc += 1

File: 08-handheldHalting/go/challenge/partOne.go
package challenge

func PartOne(instr string) int {
	instructions := parse(instr)

	var (
		acc     int
		pc      int
		visited []int
	)

	for {
		{
			var found bool
			for _, v := range visited {
				if v == pc {
					found = true
					break
				}
			}
			if found {
				return acc
			} else {
				visited = append(visited, pc)
			}
		}

		cir := instructions[pc]

		if cir.Opcode == "jmp" {
			pc += cir.Operand
		} else {
			switch cir.Opcode {
			case "acc":
				acc += cir.Operand
			case "nop":
			}

			pc += 1

		}

	}
}
